### PR TITLE
refactor(protocol-designer): begin to use getLabwareDef selector

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

Addresses a big chunk of  #3333 and #3334

OK the next few PRs from me are gonna require some explanations because there's a handful of scaffolding/shimming I'm setting up that will be removed at the end of this big move to labware defs v2 in PD but needs to exist until we arrive there.

Instead of static `getLabware` util from shared-data, this `getLabwareDef` selector is the first step towards supporting custom labware in PD and at the same time, migrating to definitions v2.

The utility functions that used to take a labware v1 `type` and then immediately call `getLabware(type)` are being rewritten to take labware v2 defs. Some of these I renamed the existing version like `blahBlahDeprecated` because a complete removal of all uses of the old fn would make this PR too huge.

* `getIsTiprackDeprecated` - used by `Labware`, `LabwareContainer`, and `HighlightableLabware` components which will be significantly rewritten soon
* `getWellsForTipsDeprecated` - used in `step-generation`, will be removed in an upcoming `step-generation` refactor pass
* `getWellSetForMultichannelDeprecated` - used by `handleFormChange` machinery, will be removed in "form stuff pass" which I am working on as the immediate follow-up to this
* `computeWellAccessDeprecated` - used in `step-generation`, used by `getWellSetForMultichannelDeprecated`, and used by some substeps stuff (so this will probably be the last of these 4 to go)

ANOTHER NOTE: I'm probably going to have to move the `labware-defs/` atom into `step-forms` reducer so that `handleFormChange` can have access to it, which I only realized after doing this work. Likely next PR I'll move it in there.

## changelog

- Replace some functions tied to v1 labware `type` string with fns that take full v2 defs
- Load labware into PD with `require.context`

## review requests

> NOTE: the shim that maps old labware to new labware is very temporary, its main purpose is to make some v2 def available for any old labware "type" so that we can replace functions that use the "type" with ones that use a v2 def. Since when we release we'll have a real migration that does this mapping correctly & robustly before the loaded file actually hits Redux, by which time the shim can be removed, this shim's mapping can be more or less arbitrary because it's only for getting us thru this brief stage of development. For example, it's mapping 200uL tipracks to 96 well plates b/c there's no 200uL def yet, and that's just something to be aware of when testing and will be resolved later in a the migration part of this once we make those defs.

**TLDR:** test this with the console open and keep an eye out for "No def in V1_NAME_TO_V2_OTID for "_______", using 96 generic for now"; be aware of the mis-mapping but it's an issue for later.

- Code review: no other consumers of shared-data or components library should be negatively affected by this PR
- Labware library `LabwareRender` component acts the same
- Component library `Labware` acts the same

PD:
- Tip position modal correctly gets depth from labware def
- Max volume warnings get max volume from labware def (eg adding liquids to deck in Initial Deck Setup; overflowing a destination well with a big transfer)
- Overall liquid tracking still works correctly across the timeline
- Overall tip racking still works correctly across the timeline, for both single and multi-channel pipettes (except for 200uL tipracks, plz ignore)
- Tip highlighting across step + substep hover works correctly, for both single and multi-channel pipettes (except for 200uL tipracks, plz ignore)
- Well selection in a step works correctly, for both single and multi-channel pipettes

### CODE REVIEW

File: protocol-designer/src/well-selection/utils.js [hunk 1]
```diff
@@ -36,29 +43,78 @@ function _getAllWellSetsForLabware(labwareName: string): ?WellSetByWell {
...
+export const getWellSetForMultichannelDeprecated = memoize(
+  _getWellSetForMultichannelDeprecated,
   (labwareName: string, well: string) => `$LABWARE:${labwareName}--WELL:${well}`
 )
 
+/** Compute all well sets for a labware type.
+ * A well set is array of 8 wells that an 8 channel pipettes can fit into,
+ * eg ['A1', 'C1', 'D1', 'E1', 'G1', 'I1', 'K1', 'M1'] is a well set in a 384 plate.
+ **/
+// TODO IMMEDIATELY - break out and memoize
```
Comment (IanLondon) [inline protocol-designer/src/well-selection/utils.js]: oops ignore this, it is broken out and memoized. I have deleted this comment in my follow-up PR to this one

File: protocol-designer/src/labware-defs/selectors.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import type { BaseState, Selector } from '../types'
+import type { DefsByLabwareId } from './types'
+import type { RootState } from './reducers'
+
+export const rootSelector = (state: BaseState): RootState => state.labwareDefs
+
+const _getLabwareDef = (
+  customDefs: DefsByLabwareId,
+  otId: string
+): any | LabwareDefinition2 => {
```
Comment (mcous) [inline protocol-designer/src/labware-defs/selectors.js]: Why is this an `any` instead of `LabwareDefinition`?

File: protocol-designer/src/step-forms/selectors/index.js [hunk 3]
```diff
@@ -58,6 +62,53 @@ export const getLabwareTypesById: Selector<LabwareTypeById> = createSelector(
     mapValues(labwareEntities, (labware: LabwareEntity) => labware.type)
 )
 
+// NOTE: this mapping should be reviewed before using it in migration later
+const V1_NAME_TO_V2_OTID = {
```
Comment (mcous) [inline protocol-designer/src/step-forms/selectors/index.js]: Do we have any tests covering this map? PRs that update/fix definitions have in the past changed the `id` (see #3289), which would break this as long as `shared-data` is our LW database

`loadName` mapping wouldn't have this same problem, but may have different issues

File: protocol-designer/src/well-selection/test/utils.test.js [hunk 4]
```diff
@@ -1,9 +1,124 @@
...
+
+// TODO Ian 2019-04-12: create representative fixtures, don't use real defs
+const fixture96Plate = require('@opentrons/shared-data/definitions2/generic_96_wellplate_380_ul.json')
+const fixture384Plate = require('@opentrons/shared-data/definitions2/corning_384_wellplate_112_ul.json')
+const fixture12Trough = require('@opentrons/shared-data/definitions2/usa_scientific_12_trough_22_ml.json')
 
 describe('getWellSetForMultichannel (integration test)', () => {
+  test('96-flat', () => {
+    const labware = fixture96Plate
+    expect(getWellSetForMultichannel(labware, 'A1')).toEqual([
```
Comment (mcous) [inline protocol-designer/src/well-selection/test/utils.test.js]: Given the TODO above about fixtures, would these maybe be better as snapshot tests on the output array?

File: shared-data/js/labwareTools/index.js [hunk 5]
```diff
@@ -101,12 +101,16 @@ export function _calculateWellCoord(
   offset: Offset,
   well: InputWell
 ): Well {
-  return {
-    ...well,
+  const coords = {
```
Comment (mcous) [inline shared-data/js/labwareTools/index.js]: Can you add a comment that these changes (that don't appear to do anything) are to support flow?

Aside: the fact that we need to make this sort of change makes me feel like we should adjust the typedefs and/or the generator functions more drastically

File: protocol-designer/src/well-selection/utils.js [hunk 6]
```diff
@@ -10,7 +15,9 @@ type WellSetByWell = Array<Array<string>>
  * A well set is array of 8 wells that an 8 channel pipettes can fit into,
  * eg ['A1', 'C1', 'D1', 'E1', 'G1', 'I1', 'K1', 'M1'] is a well set in a 384 plate.
```
Comment (b-cooper) [inline protocol-designer/src/well-selection/utils.js]: while you're here, Remove 'D1' and append 'O1' to the end

File: protocol-designer/src/well-selection/utils.js [hunk 7]
```diff
@@ -36,29 +43,78 @@ function _getAllWellSetsForLabware(labwareName: string): ?WellSetByWell {
...
-export const getWellSetForMultichannel = memoize(
-  _getWellSetForMultichannel,
+export const getWellSetForMultichannelDeprecated = memoize(
+  _getWellSetForMultichannelDeprecated,
   (labwareName: string, well: string) => `$LABWARE:${labwareName}--WELL:${well}`
 )
 
+/** Compute all well sets for a labware type.
+ * A well set is array of 8 wells that an 8 channel pipettes can fit into,
+ * eg ['A1', 'C1', 'D1', 'E1', 'G1', 'I1', 'K1', 'M1'] is a well set in a 384 plate.
```
Comment (b-cooper) [inline protocol-designer/src/well-selection/utils.js]: Same - 'D1' + 'O1'

File: protocol-designer/src/step-forms/selectors/index.js (hunk 3 above)
Comment (IanLondon) [reply protocol-designer/src/step-forms/selectors/index.js]: this is just a shim so that some v2 def data can get associated with each labware, once we use this in migration it would make sense to test that all the `otId`s actually exist to make any PRs which would remove those IDs fail. But we're not supposed to be changing ids once a def is released, right?

Comment (mcous) [reply protocol-designer/src/step-forms/selectors/index.js]: If we treat a labware definition like a software package/library (just for discussion purposes I mean), then what I've got in my head is that a load name is the unique name of the package that may have more than one release, and the ID is the identifier of a specific release/version.

Thanks for the explanation, you're TODO comment covers my concern

File: protocol-designer/src/labware-defs/selectors.js (hunk 2 above)
Comment (IanLondon) [reply protocol-designer/src/labware-defs/selectors.js]: good catch that's a leftover, deleted this in follow-up PR

File: protocol-designer/src/well-selection/utils.js (hunk 7 above)
Comment (IanLondon) [reply protocol-designer/src/well-selection/utils.js]: :eyes: updated both in next PR :+1:
